File: kirun-js/src/engine/function/system/math/Maximum.ts
import { Schema } from '../../../json/schema/Schema';
import { Event } from '../../../model/Event';
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionSignature } from '../../../model/FunctionSignature';
import { Parameter } from '../../../model/Parameter';
import { Namespaces } from '../../../namespaces/Namespaces';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { AbstractFunction } from '../../AbstractFunction';

const VALUE = 'value';

const SIGNATURE = new FunctionSignature('Maximum')
    .setNamespace(Namespaces.MATH)
    .setParameters(
        new Map([[VALUE, new Parameter(VALUE, Schema.ofNumber(VALUE)).setVariableArgument(true)]]),
    )
    .setEvents(new Map([Event.outputEventMapEntry(new Map([[VALUE, Schema.ofNumber(VALUE)]]))]));

export class Maximum extends AbstractFunction {
    public getSignature(): FunctionSignature {
        return SIGNATURE;
    }

    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        let nums: number[] = context.getArguments()?.get(VALUE);

        return new FunctionOutput([
            EventResult.outputOf(
                new Map([
                    [VALUE, nums.reduce((a, c) => ((!a && a !== 0) || c > a ? c : a))],
                ]),
            ),
        ]);
    }
}

File: kirun-py/src/kirun_py/function/system/math/maximum.py
from __future__ import annotations

from typing import List, TYPE_CHECKING

from kirun_py.function.abstract_function import AbstractFunction
from kirun_py.json.schema.schema import Schema
from kirun_py.model.event import Event
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput
from kirun_py.model.function_signature import FunctionSignature
from kirun_py.model.parameter import Parameter
from kirun_py.namespaces.namespaces import Namespaces

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters

VALUE = 'value'


class Maximum(AbstractFunction):

    def __init__(self) -> None:
        super().__init__()
        self._signature = (
            FunctionSignature('Maximum')
            .set_namespace(Namespaces.MATH)
            .set_parameters({
                VALUE: Parameter(VALUE, Schema.of_number(VALUE))
                .set_variable_argument(True),
            })
            .set_events(
                dict([Event.output_event_map_entry({VALUE: Schema.of_number(VALUE)})])
            )
        )

    def get_signature(self) -> FunctionSignature:
        return self._signature

    async def internal_execute(
        self, context: FunctionExecutionParameters
    ) -> FunctionOutput:
        nums: List[float] = context.get_arguments().get(VALUE)
        result = max(nums)
        return FunctionOutput([
            EventResult.output_of({VALUE: result})
        ])
